src/dmd/expressionsem.d

```diff
@@ -5129,12 +5129,40 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             }
             else if (sc.func)
             {
+                FuncDeclaration fd      = s.isFuncDeclaration();
+                AggregateDeclaration ag = s.isAggregateDeclaration();
+                EnumDeclaration ed      = s.isEnumDeclaration();
+                bool inserted     =  (fd || ag || ed || (v && v.isDataseg())) && sc.func.localsymtab.insert(s);
+                //  Try to insert static decl from subscope, see https://issues.dlang.org/show_bug.cgi?id=issue 3031
+                //  ---------------------------------------------------
+                //  let the func body `{ { static T decl; } { static T decl; } }`
+                //  second `decl` insertion fails.
+                //  the fix is to replace the static decl name and use the old sym name as **local** alias,
+                //  to get            `{ { static T decl; } { static T RndID; alias decl = RndID; }`
+                //  as 1) it is not possible to link against a global var that is hidden in a func body.
+                //     2) in a func body, out of order decls are not supported so we can patch the name without affecting prior code
+                if (!inserted && (sc.flags & SCOPE.subfuncscope) &&
+                    ((v && v.isDataseg()) || ed || (fd && fd.storage_class & STC.static_)))
+                {
+                    // printf("fixing 3031, on %s at line %d\n", s.toChars(), s.loc.linnum);
+                    scope OutBuffer ob;
+                    ob.printf("%s_L%d_C%d_id", s.ident.toChars(), s.loc.linnum, s.loc.charnum);
+
+                    Identifier oldID    = s.ident;
+                    s.ident             = Identifier.generateId(ob.extractSlice());
+                    AliasDeclaration ad = new AliasDeclaration(s.loc, oldID, s);
+
+                    dsymbolSemantic(s, sc);
+                    dsymbolSemantic(ad, sc);
+
+                    sc.func.localsymtab.insert(ad);
+                    sc.func.localsymtab.insert(s);
+
+                    inserted = true;
+                }
                 // https://issues.dlang.org/show_bug.cgi?id=11720
                 // include Dataseg variables
-                if ((s.isFuncDeclaration() ||
-                     s.isAggregateDeclaration() ||
-                     s.isEnumDeclaration() ||
-                     v && v.isDataseg()) && !sc.func.localsymtab.insert(s))
+                if (!inserted && (fd || ag || ed || (v && v.isDataseg())))
```

Comment:
`inserted` can be dispensed with if `if` is replaced with `else if`.

Reply:
it is required because otherwise after a successful insertion  (e.g when there's not conflict) the path of the `else if` branch is taken.